nxp_s32k/libuavcan/driver/src/s32k_libuavcan.cpp

```diff
@@ -0,0 +1,777 @@
+    volatile std::uint32_t delta = 0;
+
+    /* Disable LPIT channel 2 for loading */
+    LPIT0->CLRTEN |= LPIT_CLRTEN_CLR_T_EN_2(1);
+
+    /* Load LPIT with its maximum value */
+    LPIT0->TMR[2].TVAL = LPIT_TMR_CVAL_TMR_CUR_VAL_MASK;
+
+    /* Enable LPIT channel 2 for timeout start */
+    LPIT0->SETTEN |= LPIT_SETTEN_SET_T_EN_2(1);
+
+    /* Start of timed block */
+    while (delta < cycles_timeout)
+    {
+        /* Check if the flag has been set */
+        if (!(flagRegister & flag_Mask))
+        {
+            return libuavcan::Result::Success;
+        }
+
+        /* Get current value of delta */
+        delta = LPIT_TMR_CVAL_TMR_CUR_VAL_MASK - (LPIT0->TMR[2].CVAL);
+    }
+
+    /* If this section is reached, means timeout ocurred and return error status is returned */
+    return libuavcan::Result::Failure;
+}
+}  // END namespace S32K
+
+libuavcan::Result S32K_InterfaceGroup::messageBuffer_Transmit(std::uint_fast8_t iface_index,
+                                                              std::uint8_t      TX_MB_index,
+                                                              const FrameType&  frame)
+{
+    /* Get data length of the frame wished to be written */
+    std::uint_fast8_t payloadLength = frame.getDataLength();
+
+    /* Casting from uint8 to native uint32 for faster payload transfer to transmission message buffer */
+    std::uint32_t* native_FrameData = reinterpret_cast<std::uint32_t*>(const_cast<std::uint8_t*>(frame.data));
+
+    /* Fill up the payload's bytes, including the ones that don't add up to a full word e.g. 1,2,3,5,6,7 byte data
+     * length payloads */
+    for (std::uint8_t i = 0; i < (payloadLength >> 2) + std::min(1, (static_cast<std::uint8_t>(payloadLength) & 0x3));
+         i++)
+    {
+        /* FlexCAN natively transmits the bytes in big-endian order, in order to transmit little-endian for UAVCAN,
+         * a byte swap is required */
+        REV_BYTES_32(native_FrameData[i],
+                     S32K::FlexCAN[iface_index]->RAMn[TX_MB_index * S32K::MB_Size_Words + S32K::MB_Data_Offset + i]);
+    }
+
+    /* Fill up frame ID */
+    S32K::FlexCAN[iface_index]->RAMn[TX_MB_index * S32K::MB_Size_Words + 1] = frame.id & CAN_WMBn_ID_ID_MASK;
+
+    /* Fill up word 0 of frame and transmit it
+     * Extended Data Length       (EDL) = 1
+     * Bit Rate Switch            (BRS) = 1
+     * Error State Indicator      (ESI) = 0
+     * Message Buffer Code       (CODE) = 12 ( Transmit data frame )
+     * Substitute Remote Request  (SRR) = 0
+     * ID Extended Bit            (IDE) = 1
+     * Remote Tx Request          (RTR) = 0
+     * Data Length Code           (DLC) = frame.getdlc()
+     * Counter Time Stamp  (TIME STAMP) = 0 ( Handled by hardware )
+     */
+    S32K::FlexCAN[iface_index]->RAMn[TX_MB_index * S32K::MB_Size_Words] =
+        CAN_RAMn_DATA_BYTE_1(0x20) | CAN_WMBn_CS_DLC(frame.getDLC()) | CAN_RAMn_DATA_BYTE_0(0xCC);
```

Comment:
CAN_WMBn_CS_DLC relies on a C-style cast which can reinterpret a value. To do this safely first use a static_cast to the underlying type and then assign without a cast to uint32_t. This will ensure the compiler would complain if there was anything unsafe before we hand this over to the macro.

```
const std::uint32_t dlc = static_cast<std::underlying_type<libuavcan::media::CAN::FrameDLC>::type>(frame.getDLC());

```

Reply:
Added in https://github.com/noxuz/platform_specific_components/commit/322b5d4d13088400d4c120ad419604792ebf8a82